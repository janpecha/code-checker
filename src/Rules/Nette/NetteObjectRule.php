<?php

	declare(strict_types=1);

	namespace JP\CodeChecker\Rules\Nette;

	use JP\CodeChecker\CommitMessage;
	use JP\CodeChecker\FileContent;
	use JP\CodeChecker\PhpRule;
	use JP\CodeChecker\Reporter;


	class NetteObjectRule extends PhpRule
	{
		public function getCommitMessage(): ?CommitMessage
		{
			return new CommitMessage(
				subject: 'Nette: replaced deprecated Nette\\Object by Nette\\SmartObject'
			);
		}


		public function processContent(
			FileContent $fileContent,
			Reporter $reporter
		): void
		{
			$fileContent->findAndReplace(
				'#(class [A-Z][a-zA-Z0-9_]+)\\sextends\\s(\\\\?Nette\\\\)Object((?:\\simplements [a-zA-Z0-9_\\\\]+){0,1}\\n(\\s*){)#m',
				"$1$3\n$4\tuse $2SmartObject;\n",
				$reporter,
				'Nette: Nette\\Object replaced by Nette\\SmartObject (deprecated in v2.4.0)'
			);
		}
	}
