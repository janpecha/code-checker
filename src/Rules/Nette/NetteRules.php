<?php

	declare(strict_types=1);

	namespace JP\CodeChecker\Rules\Nette;

	use JP\CodeChecker\CheckerConfig;
	use JP\CodeChecker\PhpRule;


	class NetteRules
	{
		/**
		 * @return PhpRule[]
		 */
		public static function create(CheckerConfig $config): array
		{
			return array_merge(
				NetteUtilsRules::create($config)
			);
		}
	}
