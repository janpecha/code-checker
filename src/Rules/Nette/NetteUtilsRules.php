<?php

	declare(strict_types=1);

	namespace JP\CodeChecker\Rules\Nette;

	use JP\CodeChecker\CheckerConfig;
	use JP\CodeChecker\Rule;


	class NetteUtilsRules
	{
		/**
		 * @return Rule[]
		 */
		public static function create(CheckerConfig $config): array
		{
			$composerVersions = $config->getComposerVersions();

			if (!$composerVersions->hasPackage('nette/utils')) {
				return [];
			}

			$version = $composerVersions->getVersion('nette/utils');

			$rules = [];

			if ($version->isEqualOrGreater('2.4.0')) {
				$rules[] = new NetteObjectRule;
			}

			return $rules;
		}
	}
