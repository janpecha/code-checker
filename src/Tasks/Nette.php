<?php

	declare(strict_types=1);

	namespace JP\CodeChecker\Tasks;

	use JP\CodeChecker\ComposerVersions;
	use JP\CodeChecker\CheckerConfig;
	use Nette\CodeChecker\Result;


	class Nette
	{
		public static function configure(CheckerConfig $config, ComposerVersions $composerVersions): void
		{
			Neon::configure($config);
			Latte::configure($config, $composerVersions->getVersion('latte/latte'));
			$config->addTask([self::class, 'netteObjectFixer'], '*.php');
			$config->addTask([self::class, 'presenterMethods'], '*.php');
		}


		public static function netteObjectFixer(string &$contents, Result $result): void
		{
			Helpers::findAndReplace(
				$contents,
				$result,
				'#(class [A-Z][a-zA-Z0-9_]+)\\sextends\\s(\\\\?Nette\\\\)Object((?:\\simplements [a-zA-Z0-9_\\\\]+){0,1}\\n(\\s*){)#m',
				"$1$3\n$4\tuse $2SmartObject;\n",
				'Nette: Nette\\Object replaced by Nette\\SmartObject (deprecated in v2.4.0)'
			);
		}


		public static function presenterMethods(string &$contents, Result $result): void
		{
			Helpers::findAndReplace(
				$contents,
				$result,
				'#->isPost\\(\\)#m',
				'->isMethod(\'POST\')',
				'Nette: HTTP - method isPost() is deprecated, use isMethod(\'POST\') (deprecated in v2.4.0)'
			);
		}
	}
