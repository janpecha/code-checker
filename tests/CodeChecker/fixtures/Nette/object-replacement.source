<?php

	namespace MyApp\Model;

	use Nette;


	class Test extends \Nette\Object
	{
		private $abc;
	}


	class Test2 extends Nette\Object
	{
		private $abc;
	}


	class Test3 extends Nette\Object implements \App\ISome
	{
		private $abc;
	}
