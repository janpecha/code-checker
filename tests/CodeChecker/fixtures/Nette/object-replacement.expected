<?php

	namespace MyApp\Model;

	use Nette;


	class Test
	{
		use \Nette\SmartObject;

		private $abc;
	}


	class Test2
	{
		use Nette\SmartObject;

		private $abc;
	}


	class Test3 implements \App\ISome
	{
		use Nette\SmartObject;

		private $abc;
	}
